### PR TITLE
Add notifier for pagerduty

### PR DESCRIPTION
**Pull request by utkarsharma2**
Added notifier for pagerduty.

Example usage - 
```
BashOperator(
            task_id="mytask",
            bash_command="redis-cli ping",
            on_failure_callback=[
                send_pagerduty_notification(
                    summary="The task {{ ti.task_id }} failed",
                    severity="critical",
                    source="airflow dag_id: {{dag.dag_id}}",
                    dedup_key="{{dag.dag_id}}-{{ti.task_id}}",
                    group="{{dag.dag_id}}",
                    component="airflow",
                    class_type="Prod Data Pipeline",
                )
            ],
        )
```

### CODE REVIEW

File: airflow/providers/pagerduty/notifications/pagerduty_notifier.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (eladkal) [inline airflow/providers/pagerduty/notifications/pagerduty_notifier.py]: please rename file to `pagerdury.py`

No need for `pagerdury_` prefix

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    from airflow.providers.pagerduty.notifications.pagerduty_notifier import send_pagerduty_notification
+
+    with DAG(
+        dag_id="pagerduty_notifier",
+        schedule_interval=None,
+        start_date=datetime(2023, 1, 1),
+        catchup=False,
+        on_success_callback=[
+            send_pagerduty_notification(
+                summary="DISK at 99%",
```
Comment (ashb) [inline docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]: Lets come up with a more realistic example attributes here please

File: airflow/providers/pagerduty/notifications/pagerduty_notifier.py (hunk 1 above)
Comment (utkarsharma2) [reply airflow/providers/pagerduty/notifications/pagerduty_notifier.py]: Yup make sense, already within path `airflow/providers/pagerduty/` - changed it.

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst (hunk 2 above)
Comment (utkarsharma2) [reply docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]: @ashb Changed the example, is this example fine?

File: airflow/providers/pagerduty/notifications/pagerduty_notifier.py (hunk 1 above)
Comment (eladkal) [reply airflow/providers/pagerduty/notifications/pagerduty_notifier.py]: see https://github.com/apache/airflow/pull/31244

Comment (utkarsharma2) [reply airflow/providers/pagerduty/notifications/pagerduty_notifier.py]: Renamed 
```
airflow/providers/pagerduty/notifications/pagerduty_notifier.py -> airflow/providers/pagerduty/notifications/pagerduty.py
tests/providers/pagerduty/notifications/test_pagerduty_notifier.py -> tests/providers/pagerduty/notifications/test_pagerduty.py
```

**Review by eladkal — DISMISSED** [review]
You will need to add entry in provider.yaml (see https://github.com/apache/airflow/pull/29191 )
the entry is needed so the new notifier will be listed in the core-extensions section

**Comment by utkarsharma2** [comment]
@eladkal Added the changes in provider.yml

I'm new to codebase, just out of curiosity. what's the purpose of `provider.yml` file?

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    specific language governing permissions and limitations
+    under the License.
+
+How-to Guide for Pagerduty notifications
+========================================
+
+Introduction
+------------
+Pagerduty notifier (:class:`airflow.providers.pagerduty.notifications.pagerduty.PagerdutyNotifier`) allows users to send
+messages to a pagerduty using the various ``on_*_callbacks`` at both the DAG level and Task level
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]:
```suggestion
The Pagerduty notifier (:class:`airflow.providers.pagerduty.notifications.pagerduty.PagerdutyNotifier`) allows users to send
messages to Pagerduty using the various ``on_*_callbacks`` at both the DAG level and Task level.
```

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    from datetime import datetime
+    from airflow import DAG
+    from airflow.operators.bash import BashOperator
+    from airflow.providers.pagerduty.notifications.notifier import send_pagerduty_notification
+
+    with DAG(dag_id="pagerduty_notifier", schedule_interval=None, start_date=datetime(2023, 1, 1), catchup=False):
+        BashOperator(
+            task_id="mytask",
+            bash_command="redis-cli ping",
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]:
```suggestion
            bash_command="fail",
```

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+-------------
+
+.. code-block:: python
+
+    from datetime import datetime
+    from airflow import DAG
+    from airflow.operators.bash import BashOperator
+    from airflow.providers.pagerduty.notifications.notifier import send_pagerduty_notification
+
+    with DAG(dag_id="pagerduty_notifier", schedule_interval=None, start_date=datetime(2023, 1, 1), catchup=False):
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]:
```suggestion
    with DAG("pagerduty_notifier", start_date=datetime(2023, 1, 1)):
```

Let's keep it simple.

Also, should we have a DAG level example as well?

File: tests/providers/pagerduty/notifications/test_pagerduty.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+from airflow.providers.pagerduty.notifications.pagerduty import (
+    PagerdutyNotifier,
+    send_pagerduty_notification,
+)
+
+PAGERDUTY_API_DEFAULT_CONN_ID = PagerdutyEventsHook.default_conn_name
+
+
+@pytest.fixture(scope="module", autouse=True)
+def pagerduty_api_connections():
```
Comment (jedcunningham) [inline tests/providers/pagerduty/notifications/test_pagerduty.py]: I doubt you need this fixture. See #31268 which I just opened.

File: airflow/providers/pagerduty/notifications/pagerduty.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+try:
+    from airflow.notifications.basenotifier import BaseNotifier
+except ImportError:
+    raise AirflowOptionalProviderFeatureException(
+        "Failed to import BaseNotifier. This feature is only available in Airflow versions >= 2.6.0"
+    )
+
+from airflow.providers.pagerduty.hooks.pagerduty_events import PagerdutyEventsHook
+
+ICON_URL: str = "https://raw.githubusercontent.com/apache/airflow/2.5.0/airflow/www/static/pin_100.png"
```
Comment (jedcunningham) [inline airflow/providers/pagerduty/notifications/pagerduty.py]: Is that actually used at all?

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst (hunk 5 above)
Comment (utkarsharma2) [reply docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]: @jedcunningham Added a similar example on DAG level, please check

File: tests/providers/pagerduty/notifications/test_pagerduty.py (hunk 6 above)
Comment (utkarsharma2) [reply tests/providers/pagerduty/notifications/test_pagerduty.py]: Yes, we don't removed it.

File: airflow/providers/pagerduty/notifications/pagerduty.py (hunk 7 above)
Comment (utkarsharma2) [reply airflow/providers/pagerduty/notifications/pagerduty.py]: Nope, removed it.

File: airflow/providers/pagerduty/notifications/pagerduty.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        `href`: [Optional] URL to make the image a clickable link.
+        `alt`: [Optional] Alternative text for the image.
+    :param links: List of links to include. Each dictionary in the list accepts the following keys:
+        `href`: URL of the link to be attached.
+        `text`: [Optional] Plain text that describes the purpose of the link, and can be used as the
+        link's text.
+    :param integration_key: PagerDuty Events API token
+    :param pagerduty_events_conn_id: connection that has PagerDuty integration key in the Pagerduty
+     API token field
+    :return: PagerDuty Events API v2 response.
```
Comment (utkarsharma2) [inline airflow/providers/pagerduty/notifications/pagerduty.py]:
```suggestion
```

File: docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    from airflow import DAG
+    from airflow.operators.bash import BashOperator
+    from airflow.providers.pagerduty.notifications.notifier import send_pagerduty_notification
+
+    with DAG(
+        "pagerduty_notifier",
+        start_date=datetime(2023, 1, 1),
+        on_failure_callback=[
+            send_pagerduty_notification(
+                summary="The task {{ ti.task_id }} failed",
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-pagerduty/notifications/pagerduty_notifier_howto_guide.rst]:
```suggestion
                summary="The dag {{ dag.dag_id }} failed",
```

**Review by jedcunningham — APPROVED** [review]

**Review by ephraimbuddy — APPROVED** [review]
@eladkal can you take a look again?

**Review by eladkal — COMMENTED** [review]
Sorry for the delay!
LGTM!

Thank you @utkarsharma2

**Pull request merged**